DRA scheduler: implement filter timeout

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The intent is to catch abnormal runtimes with the generously large default timeout of 10 seconds, as discussed here:
- https://github.com/kubernetes/kubernetes/issues/131730#issuecomment-2873598287
- https://github.com/kubernetes/enhancements/pull/5333#discussion_r2103914175


#### Which issue(s) this PR fixes:

Related-to: https://github.com/kubernetes/kubernetes/issues/131730#issuecomment-2873598287, https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

We have to set up a context with the configured timeout (optional!), then ensure that both CEL evaluation and the allocation logic itself properly returns the context error. The scheduler plugin then can convert that into "unschedulable".

#### Does this PR introduce a user-facing change?

```release-note
DRA: the scheduler plugin now prevents abnormal filter runtimes by timing out after 10 seconds. This is configurable via the plugin configuration's `FilterTimeout`. Setting it to zero disables the timeout and restores the behavior of Kubernetes <= 1.33.
```

pohly committed:
DRA scheduler: add plugin configuration

The only option is the filter timeout.
The implementation of it follows in a separate commit.

pohly committed:
DRA scheduler: implement filter timeout and cancellation

The intent is to catch abnormal runtimes with the generously large default
timeout of 10 seconds.

We have to set up a context with the configured timeout (optional!), then
ensure that both CEL evaluation and the allocation logic itself properly
returns the context error. The scheduler plugin then can convert that into
"unschedulable".

The allocator and thus Filter now also check for context cancellation by the
scheduler. This happens when enough nodes have been found.

pohly commented:
/retest

Some known flakes, timeouts.

pohly committed:
DRA integration: remove unnecessary anonymous import

It's unclear why k8s.io/kubernetes/pkg/apis/resource/install needs
to be imported explicitly. Having the apiserver and scheduler ready
to be started ensures that all APIs are available.

pohly committed:
DRA integration: add test case for FilterTimeout

This covers disabling the feature via the configuration, failing to schedule
because of timeouts for all nodes, and retrying after ResourceSlice changes with
partial success (timeout for one node, success for the other).

While at it, some helper code gets improved.

carlory commented on pkg/scheduler/apis/config/validation/validation_pluginargs.go:
```diff
@@ -327,3 +327,12 @@ func ValidateNodeResourcesFitArgs(path *field.Path, args *config.NodeResourcesFi
+func ValidateDynamicResourcesArgs(path *field.Path, args *config.DynamicResourcesArgs) error {
+	var allErrs field.ErrorList
+	if args.FilterTimeout != nil && args.FilterTimeout.Duration < 0 {
+		allErrs = append(allErrs, field.Invalid(path.Child("filterTimeout"), args.FilterTimeout, "must be positive"))
```
```suggestion
		allErrs = append(allErrs, field.Invalid(path.Child("filterTimeout"), args.FilterTimeout, ""must be zero or positive"))
```

pohly committed:
scheduler: enhance and document Filter cancellation

When using context.CancelCause in the scheduler and context.Cause in plugins,
the status returned by plugins is more informative than just "context
canceled".

Context cancellation itself is not new, but many plugin authors probably
weren't aware of it because it wasn't documented.

pohly committed:
scheduler_perf: document benchstat usage

With benchstat it's easy to do before/after comparisons, but the section for
running benchmark didn't mention it at all and didn't work as shown there:
- benchmark results must be printed (FULL_LOG)
- timeout might have been too short (KUBE_TIMEOUT)
- only "short" benchmarks ran (SHORT)
- klog log output must be redirected (ARTIFACTS)

sanposhiho commented on pkg/scheduler/apis/config/validation/validation_pluginargs.go:
```diff
@@ -327,3 +327,12 @@ func ValidateNodeResourcesFitArgs(path *field.Path, args *config.NodeResourcesFi
+// ValidateDynamicResourcesArgs validates that DynamicResourcesArgs are correct.
+func ValidateDynamicResourcesArgs(path *field.Path, args *config.DynamicResourcesArgs) error {
+	var allErrs field.ErrorList
+	if args.FilterTimeout != nil && args.FilterTimeout.Duration < 0 {
```
Oh, could `Duration` be nil actually? (like people are setting -10s on `FilterTimeout` in that case?)

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -122,13 +127,23 @@ func New(ctx context.Context, plArgs runtime.Object, fh framework.Handle, fts fe
+	if !ok {
+		return nil, fmt.Errorf("got args of type %T, want *DynamiccResourcesArgs", plArgs)
+	}
+	if err := validation.ValidateDynamicResourcesArgs(nil, args); err != nil {
+		return nil, err
+	}
```
nit: so that the validation isn't triggered when the feature gate is off

```suggestion
	if fts.EnableDRASchedulerFilterTimeout {
		if err := validation.ValidateDynamicResourcesArgs(nil, args); err != nil {
			return nil, err
		}
	}
```

pohly replied:
Okay. I suppose the corresponding behavior in the apiserver is to drop disabled fields before they reach validation, i.e. invalid values are okay when the corresponding feature is off.

However, we only want to disable validation of that one field. Other fields might get added later, so we need to push the feature gates into `ValidateDynamicResourcesArgs` and do the if check there.

pohly replied:
Yes, but then I included that additional commit in the wrong branch. Thanks for double-checking! It's good to know that you are diligent about reviews despite the time constraints.

It's here now, see 8c13b2c99beb5fe5f46f18a3ae26ec141909b892.

pohly replied:
This turned out be more complicated: that commit broke `validatePluginConfig`, which called `ValidateDynamicResourcesArgs` through reflection => runtime error instead of compile error, so I hadn't noticed.

To solve this, I had to find a way to construct the struct from the `DefaultFeatureGate` also in that code. I ended up moving the construction next to the struct, which should also make future extensions more localized.

See the [updated commit](https://github.com/kubernetes/kubernetes/pull/132033/commits/b858402a370b531e7e689f5afe92e18a562dff80) for the complete solution.

sanposhiho replied:
Thanks, yes, `features.NewFeatures` makes sense.

/approve
from sig-scheduling (non API area

sanposhiho commented on staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go:
```diff
@@ -126,6 +126,9 @@ type Options struct {
 	DisableCostEstimation bool
+
+	// DisableInterrupts can be set to skip checking of the context for cancellation during evaluation.
+	DisableInterrupts bool
```
Is it always false? I didn't see where the value is set in a non-test code.

pohly replied:
It's indeed always false in production. It's useful also without the timeout feature because of the Filter cancellation by the scheduler.

Now that you mention it: perhaps `DisableInterrupts` should get removed entirely? I must have intended to make it depend on the feature gate, but then didn't.

pohly commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -216,3 +218,42 @@ type RequestedToCapacityRatioParam struct {
+	// usage of DRA. However, slow filtering can slow down Pod scheduling
+	// also for Pods not using DRA. Administators can reduce the timeout
+	// after checking the
+	// `scheduler_framework_extension_point_duration_seconds` metrics.
```
Good point. Then admins should look at `scheduler_plugin_execution_duration_seconds` first because that is what the timeout is applied to, with `scheduler_framework_extension_point_duration_seconds` providing some additional insights. Updated accordingly.

sanposhiho reviewed (approved):
I know the linter is faling, but leaving
/approve
from me anyway because I'm going to be unavailable next week.

leave /lgtm to @macsko. Though I believe it needs api-review anyway to get merged

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go:
```diff
@@ -3523,3 +3526,61 @@ func (l informerLister[T]) Get(name string) (*T, error) {
+				ctx = c
+			case "cancel":
+				c, cancel := context.WithCancel(ctx)
+				defer cancel()
```
Remove defer from here as well.

pohly replied:
Fixed.

macsko commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -551,8 +566,20 @@ func (pl *DynamicResources) Filter(ctx context.Context, cs fwk.CycleState, pod *
+		case errors.Is(err, context.DeadlineExceeded):
+			return statusUnschedulable(logger, "timed out trying to allocate devices", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()))
+		case ctx.Err() != nil:
+			return statusUnschedulable(logger, "asked by caller to stop allocating devices", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()), "cause", context.Cause(ctx))
```
This behavior is a bit inconsistent with the Filter's interface comment, where you mentioned that `context.Cause(ctx)` is passed to the status message. Here, you would only log the cause .

pohly replied:
True. Let's do as suggested and include the cause. I also added a test case. I need to rebase, so here's the incremental diff before the rebase for your review: https://github.com/kubernetes/kubernetes/compare/66b9406a7dd4de8a82f1c11769e3379f5c1b51cf..62bc3360c20b1dd54206d4e0a6414a8fb5618ec8

macsko commented:
/lgtm
/approve

nojnhuh commented on test/integration/scheduler_perf/README.md:
````diff
@@ -33,9 +33,13 @@ Currently the test suite has the following:
+make test-integration WHAT=./test/integration/scheduler_perf/... KUBE_TIMEOUT=-timeout=1h ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" FULL_LOG=y ARTIFACTS=/tmp SHORT=-short=false KUBE_TEST_ARGS='-run=^$ -benchtime=1ns -bench=BenchmarkPerfScheduling'
 ```
 
+The output can used for `benchstat` to summarize results or to do before/after
````
Could we add a link?
```suggestion
The output can used for [`benchstat`](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to summarize results or to do before/after
```

pohly replied:
Added and rebased to address conflicts with other Filter interface changes - please re-add LGTM?

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -974,6 +1000,112 @@ func TestPlugin(t *testing.T) {
+			want: want{
+				filter: perNodeResult{
+					workerNode.Name: {
+						status: framework.NewStatus(framework.UnschedulableAndUnresolvable, `asked by caller to stop allocating devices: test canceling Filter`),
```
I had fixed the merge conflicts, but hadn't noticed that I need to update some code without conflicts. Should be fixed now.

pohly commented:
/retest

kind cluster setup failed - CI issue?

pohly commented:
/test pull-kubernetes-e2e-kind

Some SIG Storage test failure.

nojnhuh reviewed (approved):
/lgtm

nojnhuh commented:
> /lgtm

Based on the diff since @macsko's review:
```
git range-diff upstream/master 2436ee7ce0d c7447201ec7
git show c7447201ec7
```

liggitt commented:
Getting oriented… do we already trace or record metrics for execution time of plugin stages so that distribution of run time and excessive runtimes could be detected even if they didn't cross a hard 10 second limit?

pohly commented:
> do we already trace or record metrics for execution time of plugin stages so that distribution of run time and excessive runtimes could be detected even if they didn't cross a hard 10 second limit?

Yes, the scheduler has those. The KEP documents them in the context of this timeout and I have a TODO item to copy that information into the 1.34 user (+ admin!) documentation.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/stable/allocator_stable.go:
```diff
@@ -678,6 +678,10 @@ func lookupAttribute(device *draapi.BasicDevice, deviceID DeviceID, attributeNam
+	if alloc.ctx.Err() != nil {
+		return false, fmt.Errorf("filter operation aborted: %w", context.Cause(alloc.ctx))
+	}
+
```
Some changes like this need to be applied to all forks of the allocator code. Changes for beta features will get applied only to `allocator_incubating.go` (thus not breaking GA!) and changes for alpha features will go into a to-be-added `allocator_experimental.go`.

pohly replied:
All of the loops end up calling `allocateOne`. I was concerned about checking the context too often, but a performance comparison with and without this didn't show any significant difference.

The most expensive operation, CEL evaluation, checks the context internally.

liggitt commented on pkg/scheduler/testing/wrappers.go:
```diff
@@ -1230,6 +1231,11 @@ func (wrapper *ResourceSliceWrapper) Device(name string, otherFields ...any) *Re
 	return wrapper
 }
 
+func (wrapper *ResourceSliceWrapper) ResourceSliceCount(count int) *ResourceSliceWrapper {
```
I don't see this called anywhere... is it?

pohly replied:
Perhaps not now, but it seemed more consistent to have it.

pohly commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -216,3 +218,45 @@ type RequestedToCapacityRatioParam struct {
+	// another scheduling attempt might succeed. If this fails repeatedly,
+	// exponential backoff slows down future attempts.
+	//
+	// The default is 10 seconds.
```
> 10 seconds per node, or 10 seconds overall?

Per `Filter` call, which is per node, and runs in parallel.

> does this backstop timeout config really belong at the individual plugin level

I had discussed this with SIG Scheduling and the consensus was to start initially like this at the plugin level and perhaps later do something generic. Even if we add something generic, being able to configure this particular plugin differently probably would still be useful.

pohly commented on pkg/scheduler/apis/config/validation/validation_pluginargs.go:
```diff
@@ -327,3 +328,12 @@ func ValidateNodeResourcesFitArgs(path *field.Path, args *config.NodeResourcesFi
+// ValidateDynamicResourcesArgs validates that DynamicResourcesArgs are correct.
+func ValidateDynamicResourcesArgs(path *field.Path, args *config.DynamicResourcesArgs, fts feature.Features) error {
+	var allErrs field.ErrorList
+	if fts.EnableDRASchedulerFilterTimeout && args.FilterTimeout != nil && args.FilterTimeout.Duration < 0 {
```
This approach here is consistent with how disabled fields are handled in the REST API (= silently ignored).

I agree that this isn't a good user experience, so I'm fine with being more strict during config validation and error out when a disabled option is set.

I don't know how this is handled elsewhere in the scheduler. For the logging configuration we are more strict and error when trying to configure something that is disabled.

pohly commented on pkg/scheduler/apis/config/validation/validation.go:
```diff
@@ -156,6 +157,9 @@ func validatePluginConfig(path *field.Path, apiVersion string, profile *config.K
 		"PodTopologySpread":               ValidatePodTopologySpreadArgs,
 		"VolumeBinding":                   ValidateVolumeBindingArgs,
+		"DynamicResources": func(path *field.Path, args *config.DynamicResourcesArgs) error {
+			return ValidateDynamicResourcesArgs(path, args, feature.NewFeatures())
```
`feature.NewFeatures()` retrieves the user-configured feature gates. Perhaps `FromFeatureGate()` with the feature gate instance being passed is clearer?

pohly commented on pkg/scheduler/apis/config/v1/defaults.go:
```diff
@@ -242,3 +243,9 @@ func SetDefaults_NodeResourcesFitArgs(obj *configv1.NodeResourcesFitArgs) {
+
+func SetDefaults_DynamicResourcesArgs(obj *configv1.DynamicResourcesArgs) {
+	if obj.FilterTimeout == nil {
+		obj.FilterTimeout = &metav1.Duration{Duration: configv1.DynamicResourcesFilterTimeoutDefault}
```
Hmm, probably.

pohly commented on pkg/scheduler/framework/plugins/feature/feature.go:
```diff
@@ -36,3 +42,25 @@ type Features struct {
 }
+
+// NewFeatures copies the current state of the feature gates into the struct.
+func NewFeatures() Features {
```
Ack, I arrived at the some conclusion when I saw your earlier question. I'm fine with your proposed name, will switch to that.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -108,6 +111,8 @@ type DynamicResources struct {
 	enablePartitionableDevices bool
 	enableDeviceTaints         bool
+	enableFilterTimeout        bool
+	filterTimeout              metav1.Duration
```
Ack.

liggitt commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -551,8 +566,20 @@ func (pl *DynamicResources) Filter(ctx context.Context, cs fwk.CycleState, pod *
+		case errors.Is(err, context.DeadlineExceeded):
+			return statusUnschedulable(logger, "timed out trying to allocate devices", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()))
+		case ctx.Err() != nil:
+			return statusUnschedulable(logger, fmt.Sprintf("asked by caller to stop allocating devices: %v", context.Cause(ctx)), "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()))
```
ok

pohly commented on test/integration/.import-restrictions:
```diff
@@ -1,5 +1,17 @@
+  #
+  # At least the following packages follow that approach.
+  # More may be added as needed.
+  - selectorRegexp: k8s[.]io/kubernetes/test/e2e/framework/(pod|daemonset)
```
Yes, eventually I want to get to a point where integration and E2E testing can share the same helper functions. `k8s.io/kubernetes/test/utils/ktesting` provides the necessary abstraction around `go test` and Ginkgo. But this is an ongoing effort and not something that can be done quickly.

In the meantime, allowing "well written" helper packages seemed acceptable to me. I think I understand the reason for forbidding the usage earlier (ginkgo.Fail will abort the integration test when an assertion fails); it does not apply to the pod and daemonset helpers.

pohly committed:
DRA scheduler: add DRASchedulerFilterTimeout feature gate

Initializing the scheduler Features struct will be needed in different places,
therefore NewSchedulerFeaturesFromGates gets introduced. Besides, having it
next to the struct makes it easier to add new features.

The DRASchedulerFilterTimeout feature gate simplifies disabling the timeout
because setting a feature gate is often easier than modifying the scheduler
configuration with a zero timeout value.

The timeout and feature gate are new. The gate starts as beta and enabled by
default, which is consistent with the "smaller changes with low enough risk
that still may need to be disabled..." guideline.

liggitt commented:
Filter config plumbing and validation looks good to me. Build looks like it's failing. I didn't review the implementation much at all, relying on area owners to review that.

Ping me when the rest of the PR is approved and I'll stamp for the API config bits

liggitt commented:
thanks, missed that in the collapsed github comments

/lgtm
/approve
